File: Numeros Pseudo-Aleatorios/Linear-Feedback Shift Register/lfsr.cs
/*
   PRNG .......: Linear-Feedback Shift Register (LFSR)
   Ano ........: 1965
   Autor ......: R. C. Tausworthe
*/

using System;

namespace LinearFeedbackShiftRegister
{

    class Program
    {
      
        static double Lfsr(ref uint state)
        {
            state = (state >> 1) ^ (0xB4BCD35C * (state & 1));
            return (double)state / 4294967296.0;
        }

        static void Main(string[] args)
        {
          
            uint Xn1, Xn2;
            int i;

            Xn1 = 0x1234;
            Xn2 = (uint)DateTime.Now.Ticks;

            Console.WriteLine("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)");
            Console.WriteLine("PSEUDORANDOM NUMBER GENERATOR (PRNG)");
            Console.WriteLine("PADRAO: Linear-Feedback Shift Register");
            Console.WriteLine("==========================================\n");
            Console.WriteLine("    Semente: 1234        Semente: TIME");
            Console.WriteLine("    -------------        -------------\n");

            for (i = 1; i <= 10; i++)
            {
                Console.WriteLine("{0,17:F11}        {1,13:F11}", Lfsr(ref Xn1), Lfsr(ref Xn2));
            }
          
        }
      
    }
}

File: Numeros Pseudo-Aleatorios/Linear-Feedback Shift Register/lfsr.pas
{
 PRNG .......: Linear-Feedback Shift Register (LFSR)
 Ano ........: 1965
 Autor ......: R. C. Tausworthe
}

program LinearFeedbackShiftRegister;

uses
  SysUtils, DateUtils;
  
function lfsr(var state: Cardinal): Double;
begin
  state := (state shr 1) xor ($B4BCD35C * (state and 1));
  lfsr := state / 4294967296.0;
end;

var
  Xn1, Xn2: Cardinal;
  i: Integer;

begin

  Xn1 := $1234; 
  Xn2 := LongWord(DateTimeToUnix(Now)); 

  writeln('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  writeln('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  writeln('PADRAO: Linear-Feedback Shift Register');
  writeln('==========================================');
  writeln;
  writeln('    Semente: 1234        Semente: TIME');
  writeln('    -------------        -------------');

  for i := 1 to 10 do
    writeln(lfsr(Xn1):17:11, '        ', lfsr(Xn2):13:11);
    
end.
